### PR TITLE
#33: Updated defaults to kafka 0.9

### PR DESCRIPTION
**Pull request by bbaugher**

### CODE REVIEW

File: recipes/default.rb [hunk 1]
```diff
@@ -80,13 +80,15 @@
...
 
 # Ensure the Kafka broker log (kafka data) directories exist
-node["kafka"]["server.properties"]["log.dirs"].split(",").each do |log_dir|
-  directory log_dir do
-    action :create
-    owner node["kafka"]["user"]
-    group node["kafka"]["group"]
-    mode 00700
-    recursive true
+if node["kafka"]["server.properties"].has_key? 'log.dirs'
```
Comment (bbaugher) [inline recipes/default.rb]: Kafka now has 2 log directory properties (log.dirs and log.dir). If log.dirs is not set it uses log.dir which has a default of /tmp/kafka

File: test/integration/default/serverspec/kafka_offset_monitor_spec.rb [hunk 2]
```diff
@@ -42,17 +42,18 @@
 
   it 'should be able to get monitoring details page for a consumer group' do
     # Generate a random topic and group
-    topicName = "testNewTopic_" + rand(100000).to_s
-    groupName = "testNewGroup_" + rand(100000).to_s
-      
+    topicName = "testNewTopic-" + rand(100000).to_s
+    groupName = "testNewGroup-" + rand(100000).to_s
```
Comment (bbaugher) [inline test/integration/default/serverspec/kafka_offset_monitor_spec.rb]: Kafka 0.9 produces warning about topics with _ or .

File: test/integration/default/serverspec/kafka_spec.rb [hunk 3]
```diff
@@ -118,22 +101,20 @@
...
     # We also run this as a background process so we can also start the producer
-    Kernel.system "/opt/kafka/bin/kafka-console-consumer.sh --zookeeper localhost:2181 --whitelist #{topic} >> /tmp/consumer.out 2>&1 &"
+    Kernel.system "/opt/kafka/bin/kafka-verifiable-consumer.sh --broker-list localhost:9092 --topic #{topic} --group-id #{group} >> /tmp/consumer.out 2>&1 &"
+
+    # Give consumer 5s to start up
+    sleep 5
 
     # The producer is a command that allows a user to write input to the console as 'messages' to the topic, separated by new line characters
     # In this case we run the command and write the same message several times over 5s in an attempt to ensure the consumer saw the message
-    IO.popen("/opt/kafka/bin/kafka-console-producer.sh --topic #{topic} --broker-list localhost:6667 2> /dev/null", mode='r+') do |io|
```
Comment (bbaugher) [inline test/integration/default/serverspec/kafka_spec.rb]: This was hanging for me. Switched to these new commands which seemed simpler/easier to use

File: test/integration/default/serverspec/kafka_spec.rb [hunk 4]
```diff
@@ -46,28 +46,11 @@
...
-    it { should be_grouped_into 'kafka' }
-  end
-end
-
 describe file('/etc/kafka') do
   it { should be_directory }
   it { should be_linked_to '/opt/kafka/config' }
 end
 
-['/var/kafka/data1', '/var/kafka/data2', '/var/kafka/data3'].each do |log_dir|
```
Comment (bbaugher) [inline test/integration/default/serverspec/kafka_spec.rb]: Moved these tests into unit tests

**Comment by mkwhitacre** [comment]
If we know the offset monitor isn't going to work with the default should we add some warning/logging/error out to this effect so consumers know as well?

File: attributes/default.rb [hunk 5]
```diff
@@ -5,8 +5,8 @@
 default["kafka"]["user"] = "kafka"
 default["kafka"]["group"] = "kafka"
 
-default["kafka"]["scala_version"] = "2.9.2"
-default["kafka"]["version"] = "0.8.1.1"
+default["kafka"]["scala_version"] = "2.11"
+default["kafka"]["version"] = "0.9.0.0"
```
Comment (noslowerdna) [inline attributes/default.rb]: Update README @ https://github.com/cerner/cerner_kafka#how-to-specify-where-to-download-kafka-from-and-which-version-to-install

File: attributes/default.rb [hunk 6]
```diff
@@ -31,7 +31,7 @@
 default["kafka"]["logrotate"]["frequency"] = 'daily'
 default["kafka"]["logrotate"]["rotate"] = 5
 
-default["kafka"]["server.properties"]["log.dirs"] = "/tmp/kafka-logs"
+default["kafka"]["server.properties"] = {}
```
Comment (noslowerdna) [inline attributes/default.rb]: Update in README

**Comment by noslowerdna** [comment]
Maybe also bump default offset monitor version to latest (0.2.1)

**Comment by noslowerdna** [comment]
In README might link to http://kafka.apache.org/documentation.html#upgrade_9, as it's critical to temporarily set inter.broker.protocol.version=0.8.2.X for a rolling upgrade.

**Comment by mkwhitacre** [comment]
Changing the default seems like a major version bump.

File: attributes/default.rb (hunk 6 above)
Comment (mkwhitacre) [inline attributes/default.rb]: +1 any attributes that had meaning but are now meaningless or need to be set by consumers we should call out as part of an upgrade guide.

**Comment by bbaugher** [comment]
We already did a major version bump when we remove a lot of the default attributes (https://github.com/cerner/cerner_kafka/commit/5674e21e519d65c26b566d108f8f446ec1f115a1).

**Comment by bbaugher** [comment]
> If we know the offset monitor isn't going to work with the default should we add some warning/logging/error out to this effect so consumers know as well?

Are you referring to the switch to offset storage in kafka? It looks like by default its still zookeeper ([doc](http://kafka.apache.org/documentation.html#consumerconfigs) look for `offsets.storage`).

File: attributes/default.rb (hunk 6 above)
Comment (bbaugher) [inline attributes/default.rb]: Done, https://github.com/cerner/cerner_kafka/commit/c0955cc46b0384d0bdf0c3b3f0f17404edf7a51b

File: attributes/default.rb (hunk 5 above)
Comment (bbaugher) [inline attributes/default.rb]: Done, https://github.com/cerner/cerner_kafka/commit/c0955cc46b0384d0bdf0c3b3f0f17404edf7a51b

**Comment by bbaugher** [comment]
Addressed all comments with this, https://github.com/cerner/cerner_kafka/commit/c0955cc46b0384d0bdf0c3b3f0f17404edf7a51b

**Comment by bbaugher** [comment]
Added upgrade guide to call out what we changed and was non-passive or what they should look out for, https://github.com/cerner/cerner_kafka/commit/f507c8014874a46acb4f1a15cc2daba9244c0d62

**Comment by mkwhitacre** [comment]
+1

**Comment by noslowerdna** [comment]
+1

**Comment by bbaugher** [comment]
Merging with enough +1s

**Pull request merged**